Address feedback on SSLv2 ClientHello processing

mattcaswell opened the pull request:
Feedback on the previous SSLv2 ClientHello processing fix was that it
breaks layering by reading init_num in the record layer. It also does not
detect if there was a previous non-fatal warning.

This is an alternative approach that directly tracks in the record layer
whether this is the first record.

GitHub Issue #1298

mattcaswell commented:
Ping @davidben. Thoughts?

davidben commented on ssl/record/ssl3_record.c:
```diff
@@ -162,15 +162,14 @@ int ssl3_get_record(SSL *s)
+             * means s->rlayer.is_first_record is true. Probably this is
+             * sufficient in itself instead of s->first_packet, but I am
+             * cautious. We check s->read_hash and s->enc_read_ctx to ensure
+             * this does not apply during renegotiation.
```
(I believe `read_hash`/`enc_read_ctx` are indeed redundant, yeah. Likewise with the `s->first_packet` check. Though I'm not familiar with the 1.1.0 record-layer code so perhaps I'm missing something.)

This is a remarkably long-winded comment that, now that the logic is pretty straightforward, doesn't seem to add much. I'd probably just have written:

```
/*
 * The first record received by the server may be a V2ClientHello.
 */
if (s->server && RECORD_LAYER_is_first_record(&s->rlayer)
        && (p[0] & 0x80) && (p[2] == SSL2_MT_CLIENT_HELLO)) {
```

And maybe stuck the redundant checks in as `assert`s if they seemed interesting enough.

davidben commented:
No tests? ;-)

kaduk commented on ssl/record/record_locl.h:
```diff
@@ -31,6 +31,8 @@
                                                 ((rl)->empty_record_count = 0)
 #define RECORD_LAYER_get_empty_record_count(rl) ((rl)->empty_record_count)
+#define RECORD_LAYER_is_first_record(rl)        ((rl)->is_first_record)
+#define RECORD_LAYER_set_first_record(rl, val)  ((rl)->is_first_record = (val))
```
If is_first_record is really just a boolean, you may not need the vull versatility of setting a specific vlaue; just a RECORD_LAYER_clear_first_record() call would be sufficient.

mattcaswell commented:
I've added three new commits to this:

- One addresses the feedback received so far
- One implements a set of tests for V2ClientHello processing
- One fixes a bug where we didn't send an alert if we received a non-initial record with the wrong version discovered as a result of the above tests.

mattcaswell commented:
Needs a team member review.

davidben commented:
(Non-team-member lgtm. But I don't really know perl, so I just glossed over those bits.)

kaduk commented on test/recipes/70-test_sslrecords.t:
```diff
@@ -62,6 +62,52 @@ $proxy->filter(\&add_frag_alert_filter);
+    ALERT_BEFORE_SSLV2 => 4
+};
+#Test 5: Inject an SSLv2 style record format for a TLSv1.2 ClientHello
+#my $sslv2testtype = TLSV1_2_IN_SSLV2;
```
This line is unneeded.

mattcaswell replied:
Fixed

kaduk commented on test/recipes/70-test_sslrecords.t:
```diff
@@ -131,10 +179,158 @@ sub add_frag_alert_filter
+    }
+
+    if ($sslv2testtype == ALERT_BEFORE_SSLV2
+            || $sslv2testtype == TLSV1_2_IN_SSLV2
```
I'll assume that the perl style guide says this is the correct indentation for continuation lines in conditionals.

kaduk commented on test/recipes/70-test_sslrecords.t:
```diff
@@ -131,10 +179,158 @@ sub add_frag_alert_filter
+        my $frag2 = substr $clienthello, 6, 32;
+        my $frag3 = substr $clienthello, 38;
+
+        my $fraglen = length $frag1;
```
It's not really clear that the fraglen local buys very much.

mattcaswell replied:
Just removes the need to use length 3 times - not a big deal, but not a big cost either.

kaduk commented:
Another non-team-member lgtm, modulo the commented duplicate line, and I only sort-of looked at the perl.

mattcaswell commented:
Duplicate line removed. So two non-team member reviews done - still need another team review!

mattcaswell commented:
Ping any team member!!

mattcaswell commented:
Rebased this.

t-j-h commented:
+1 - code looks much clearer to me at least

mattcaswell commented:
Merged. Thanks @t-j-h.
